### PR TITLE
test: Support tracking gas refunds

### PR DESCRIPTION
**Pull request by axic**
Depends on #13532.

### CODE REVIEW

File: test/ExecutionFramework.cpp [hunk 1]
```diff
@@ -192,14 +192,21 @@ void ExecutionFramework::sendMessage(bytes const& _data, bool _isCreation, u256
 	if (_isCreation)
 		m_contractAddress = EVMHost::convertFromEVMC(result.create_address);
 
-	m_gasUsed = InitialGas - result.gas_left;
+	unsigned const refundRatio = (m_evmVersion >= langutil::EVMVersion::london() ? 5 : 2);
+	auto const gasUsed = InitialGas - result.gas_left;
```
Comment (axic) [inline test/ExecutionFramework.cpp]: Could also use
```suggest
	auto const gasUsed = message.gas - result.gas_left;
```

**Comment by nikola-matic** [comment]
`soltest` steps are failing, I'm assuming you didn't update all of the tests?

**Comment by axic** [comment]
> soltest steps are failing, I'm assuming you didn't update all of the tests?

Yes, it takes like 2 hrs to run it locally, and will need to rebase this anyhow once the dependencies are merged.

**Comment by nikola-matic** [comment]
> > soltest steps are failing, I'm assuming you didn't update all of the tests?
> 
> Yes, it takes like 2 hrs to run it locally, and will need to rebase this anyhow once the dependencies are merged.

Ah neat, so I see that @cameel approved https://github.com/ethereum/solidity/pull/13537, which means you can resolve conflicts on [this one](https://github.com/ethereum/solidity/pull/13532) so we can merge it, and then you can rebase this one so we can merge it, and so on and so forth? :D

**Comment by nikola-matic** [comment]
@axic, dependencies are merged, so is this good to go?

**Comment by axic** [comment]
Good to go from my perspective.

File: test/ExecutionFramework.cpp [hunk 2]
```diff
@@ -192,14 +192,21 @@ void ExecutionFramework::sendMessage(bytes const& _data, bool _isCreation, u256
...
+
+	m_gasUsed = gasUsed - gasRefund;
 	m_transactionSuccessful = (result.status_code == EVMC_SUCCESS);
 
 	if (m_showMessages)
 	{
 		cout << " out:     " << util::toHex(m_output) << endl;
 		cout << " result: " << static_cast<size_t>(result.status_code) << endl;
 		cout << " gas used: " << m_gasUsed.str() << endl;
+		cout << " gas used (without refund): " << gasUsed.str() << endl;
```
Comment (nikola-matic) [inline test/ExecutionFramework.cpp]: Are these message output anywhere in the tests? I see that gas values for a lot of the tests have changed, but I'm not sure where these messages can be seen.

Comment (axic) [reply test/ExecutionFramework.cpp]: You can display them using `--show-messages` for debugging.

File: test/ExecutionFramework.cpp [hunk 3]
```diff
@@ -192,14 +192,21 @@ void ExecutionFramework::sendMessage(bytes const& _data, bool _isCreation, u256
...
+		cout << " gas refund (total): " << result.gas_refund << endl;
+		cout << " gas refund (bound): " << gasRefund.str() << endl;
```
Comment (cameel) [inline test/ExecutionFramework.cpp]: Were these meant to be aligned for readability? The leftover whitespace for `out:` would suggest that:
```suggestion
		cout << " out:                       " << util::toHex(m_output) << endl;
		cout << " result:                    " << static_cast<size_t>(result.status_code) << endl;
		cout << " gas used:                  " << m_gasUsed.str() << endl;
		cout << " gas used (without refund): " << gasUsed.str() << endl;
		cout << " gas refund (total):        " << result.gas_refund << endl;
		cout << " gas refund (bound):        " << gasRefund.str() << endl;
```

File: test/ExecutionFramework.cpp (hunk 2 above)
Comment (cameel) [inline test/ExecutionFramework.cpp]: Initially I wanted to say you're using the same var in both cases and only then noticed that tere's actually a difference. I'd make the naming more precise:
```suggestion
		cout << " gas used: " << m_gasUsed.str() << endl;
		cout << " gas used (without refund): " << gasUsedNoRefund.str() << endl;
```

**Review by cameel — DISMISSED** [review]
I have small stylistic suggestions but they're pretty minor and overall this looks fine so approving already. Feel free to ignore them if it's too much bother to rerun the test suite here.

File: test/ExecutionFramework.cpp (hunk 3 above)
Comment (axic) [reply test/ExecutionFramework.cpp]: Changed.

File: test/ExecutionFramework.cpp (hunk 2 above)
Comment (axic) [reply test/ExecutionFramework.cpp]: Changed.

**Review by nikola-matic — DISMISSED** [review]

**Comment by axic** [comment]
Needed to update tests due to rebasing over the optimizor changes which were merged.

**Review by nikola-matic — APPROVED** [review]

**Pull request merged**